src/common/mfu_daos.c

```diff
@@ -4075,122 +4199,138 @@ static int cont_deserialize_keys(struct hdf5_args *hdf5,
-                  hdf5->akey_data[ak_off + k].akey_val.len);
-
-            /* read record data for each akey */
-            index = ak_off + k;
-            len = snprintf(NULL, 0, "%lu", index);
-            char dset_name[len + 1];    
-            snprintf(dset_name, len + 1, "%lu", index);
-            hdf5->rx_dset = H5Dopen(hdf5->file, dset_name,
-                                    H5P_DEFAULT);
-            if (hdf5->rx_dset < 0) {
-                MFU_LOG(MFU_LOG_ERR, "failed to read rx_dset");
-                rc = 1;
-                goto out;
-            }
-            hdf5->rx_dspace = H5Dget_space(hdf5->rx_dset);
-            if (hdf5->rx_dspace < 0) {
-                MFU_LOG(MFU_LOG_ERR, "failed to read rx_dspace");
-                rc = 1;
-                goto out;
-            }
-            hdf5->rx_dtype = H5Dget_type(hdf5->rx_dset);
-            if (hdf5->rx_dtype < 0) {
-                MFU_LOG(MFU_LOG_ERR, "failed to read rx_dtype");
-                rc = 1;
-                goto out;
-            }
-            hdf5->plist = H5Dget_create_plist(hdf5->rx_dset);
-            if (hdf5->plist < 0) {
-                MFU_LOG(MFU_LOG_ERR, "failed to get plist");
-                rc = 1;
-                goto out;
-            }
-            rx_ndims = H5Sget_simple_extent_dims(hdf5->rx_dspace,
-                                                 hdf5->rx_dims,
-                                                 NULL);
-            if (rx_ndims < 0) {
-                MFU_LOG(MFU_LOG_ERR, "failed to get rx_ndims");
-                rc = 1;
-                goto out;
-            }
-            num_attrs = H5Aget_num_attrs(hdf5->rx_dset);
-            if (num_attrs < 0) {
-                MFU_LOG(MFU_LOG_ERR, "failed to get num attrs");
-                rc = 1;
+        /* if rec_kv_val.len != 0 then skip akey iteration, we can
+         * write data back into DAOS using the daos_kv.h API using just
+         * oid, dkey, and key value (stored in dkey dataset) */
+
+        /* run daos_kv_put on rec_kv_val (dkey val) and key (dkey) */
+        /* skip akey iteration for DAOS_OF_KV_FLAT objects */
+        if (hdf5->dkey_data[*dk_off + j].rec_kv_val.len > 0) {
+            daos_size_t kv_single_size = 0;
+            kv_single_size = hdf5->dkey_data[*dk_off + j].rec_kv_val.len;
+            rc = daos_kv_put(*oh, DAOS_TX_NONE, 0, dkey, kv_single_size,
+                             hdf5->dkey_data[*dk_off + j].rec_kv_val.p, NULL);
+            if (rc != 0) {
+                MFU_LOG(MFU_LOG_ERR, "failed to put kv object: "DF_RC, DP_RC(rc));
                 goto out;
             }
-            if (num_attrs > 0) {
-                rc = cont_deserialize_recx(hdf5, oh, diov,
-                                           num_attrs, ak_off, k, stats);
-                if (rc != 0) {
-                    MFU_LOG(MFU_LOG_ERR, "failed to deserialize recx");
+            stats->bytes_written += kv_single_size;
+        } else {
```

Comment:
@dsikich Instead of the large if/else blocks, I wonder if we could just abstract each to a function.

Reply:
@daltonbohning yeah, this function is becoming rather large. I can move them to their own functions.